SharedPackages/VPN/Sources/VPN/PacketTunnelProvider.swift

```diff
@@ -185,8 +189,39 @@ open class PacketTunnelProvider: NEPacketTunnelProvider {
 
     // MARK: - WireGuard
 
+    private lazy var wireGuardAdapterEventMapper: EventMapping<WireGuardAdapterEvent> = .init(mapping: { [weak self] event, _, _, _ in
+        guard let self else { return }
+
+        switch event {
+        case .endTemporaryShutdownStateAttemptFailure(let error):
+            Logger.networkProtection.error("Adapter failed to exit temporary shutdown: \(error.localizedDescription)")
+            self.providerEvents.fire(.adapterEndTemporaryShutdownStateAttemptFailure(error))
+        case .endTemporaryShutdownStateRecoveryFailure(let error):
+            Logger.networkProtection.error("Adapter recovery from temporary shutdown failed: \(error.localizedDescription)")
+            self.providerEvents.fire(.adapterEndTemporaryShutdownStateRecoveryFailure(error))
+        case .endTemporaryShutdownStateRecoverySuccess:
+            Logger.networkProtection.log("Adapter recovery from temporary shutdown succeeded")
+            self.providerEvents.fire(.adapterEndTemporaryShutdownStateRecoverySuccess)
+        }
+
+        if settings.showDebugVPNEventNotifications {
+            let notificationText: String
+
+            switch event {
+            case .endTemporaryShutdownStateAttemptFailure(let error):
+                notificationText = "VPN failed to end temporary shutdown: \(error.localizedDescription)"
+            case .endTemporaryShutdownStateRecoveryFailure(let error):
+                notificationText = "VPN failed to recover from extended temporary shutdown: \(error.localizedDescription)"
+            case .endTemporaryShutdownStateRecoverySuccess:
+                notificationText = "VPN recovered after extended temporary shutdown"
+            }
+
+            notificationsPresenter.showDebugEventNotification(message: notificationText)
+        }
+    })
```

Comment:
I know we haven't been doing this much, but it may be worth considering if we should abstract some of this from `PacketTunnelProvider`.

Would you be open to move this closure / logic inside a `WireguardEventHandler` that we initialize injecting the necessary dependencies for the handling?

I think that might eventually come in handy for testing and improving readability in this huge class.